cmd/tailscale/cli/ping.go

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package cli
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"log"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/peterbourgon/ff/v2/ffcli"
+	"tailscale.com/ipn"
+	"tailscale.com/ipn/ipnstate"
+)
+
+var pingCmd = &ffcli.Command{
+	Name:       "ping",
+	ShortUsage: "ping <hostname-or-IP>",
+	ShortHelp:  "Ping a host at the Tailscale layer, see how it routed",
+	LongHelp: strings.TrimSpace(`
+
+The 'tailscale ping' command pings a peer node at the Tailscale layer
+and reports which route it took for each response. The first ping or
+so will likely go over DERP (Tailscale's TCP relay protocol) while NAT
+traversal finds a direct path through.
+
+If 'tailscale ping' works but a normal ping does not, that means one
+side's operating system firewall is blocking packets; 'tailscale ping'
+does not inject packets into either side's TUN devices.
+
+By default, 'tailscale ping' stops after 10 pings or once a direct
+(non-DERP) path has been established.
+
+The provided hostname or IP must be for a Tailscale IP
+(e.g. 100.x.y.z) or an subnet IP advertised by a Tailscale relay node.
```

Comment:
s/an/a/

this is slightly awkward as written, since the two options for which the provided hostname or IP must be are both IP addresses. Could add "or a hostname that resolves to such an IP address". But I suggest instead you just delete this entire paragraph. And if it doesn't already do so, have `tailscale ping` emit a useful error message and exit when bad input is provided.

Reply:
Tailscale ping doesn't (or rather: shouldn't necessarily) have the netmap, so it won't know a resolved hostname is or is not behind a subnet router.

Reply:
Updated the text.